### PR TITLE
Sot 49 dark

### PR DESCRIPTION
**Pull request by zk-steve**
# Check List

- [x] Did you check your grammar / reader proof your document using ChatGPT?
- [x] Don't forget to squash commits into meaningful chunks before merging
- [x] After merging this, please update your task status!
- [x] Did you verify that your math formulas are rendered correctly?

### CODE REVIEW

File: docs/dark.md [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+![](attachments/dark.png)
+
+## Integer Commitment Scheme
+
+DARK uses Damgard-Fujisaki integer [commitment scheme](../terms/commitment_scheme.md), you can read the full paper of it
+[here](https://citeseerx.ist.psu.edu/document?repid=rep1&type=pdf&doi=fdfdb6df844a02ec1d783e43899a179920f13dac). This scheme is
+**additive homomorphic** and **multiplicative homomorphic**. That is, we can prove that $a + b = c$ and $a \cdot b = c$ using $com(a),
+com(b), com(c)$.
+
+For example, prover has some values $x_1, x_2, x_3, x_{4}$ and wants to prove that $x_4 = x_1x_2+x_3$.
```
Comment (zk-steve) [inline docs/dark.md:17]:
```suggestion
For example, a prover has some values $x_1, x_2, x_3, x_{4}$ and wants to prove that $x_4 = x_1x_2+x_3$.
```

File: docs/dark.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+- Prover calculates $x_5=x_1x_2$ and sends commitments of $x_1,x_2,x_3,x_4,x_5$ to verifier.
```
Comment (zk-steve) [inline docs/dark.md:19]:
```suggestion
- The prover calculates $x_5=x_1x_2$ and sends commitments of $x_1,x_2,x_3,x_4,x_5$ to verifier.
```

File: docs/dark.md [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+- Verifier checks that $x_5=x_1x_2$ and $x_4=x_3+x_5$ using their commitments.
```
Comment (zk-steve) [inline docs/dark.md:20]:
```suggestion
- The verifier checks that $x_5=x_1x_2$ and $x_4=x_3+x_5$ using their commitments.
```

File: docs/dark.md [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+## DARK
+
+### Main Idea
+
+First, you should read the definitions
```
Comment (zk-steve) [inline docs/dark.md:26]:
```suggestion
```

File: docs/dark.md [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+of [diophantine equation](../terms/diophantine_equation.md), [diophantine set](../terms/diophantine_set.md).
```
Comment (zk-steve) [inline docs/dark.md:27]:
```suggestion
```

File: docs/dark.md [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+The main idea of DARK is to prove a constraint with $\mu$ as the witness, prover needs to find an auxiliary witness $\omega$, where
+$R_S(\mu, \omega) = 0$ and then uses the integer commitment scheme above to commit to elements of $R_S$. For example, to prove $\mu
```
Comment (zk-steve) [inline docs/dark.md:30]:
```suggestion
the [diophantine equation](../terms/diophantine_equation.md) $R_S(\mu, \omega) = 0$ and then uses the integer commitment scheme above to commit to elements of $R_S$. For example, to prove $\mu
```

File: docs/dark.md [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+\geq 0$, prover
+finds $\omega_{1}, \omega_{2}, \omega_{3}, \omega_{4} \ s.t. \ R_S(\mu,\omega)=\mu - (\omega_1^2 + \omega_2^2 + \omega_3^2 +
+\omega_4^2)$ (see the
+non-negative proof below for more details). Then prover sends $com(\mu), com(\omega_1),com(\omega_2),com(\omega_3),com(\omega_4)$ to
```
Comment (zk-steve) [inline docs/dark.md:34]:
```suggestion
non-negative proof below for more details). Then the prover sends $com(\mu), com(\omega_1),com(\omega_2),com(\omega_3),com(\omega_4)$ to
```

File: docs/dark.md [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+verifier. Verifier checks $com(\mu)-(com(\omega_1)^2+com(\omega_2)^2+com(\omega_3)^2+com(\omega_4)^2)=com(0)$
```
Comment (zk-steve) [inline docs/dark.md:35]:
```suggestion
the verifier. Verifier checks $com(\mu)-(com(\omega_1)^2+com(\omega_2)^2+com(\omega_3)^2+com(\omega_4)^2)=com(0)$
```

File: docs/dark.md (hunk 4 above)
Comment (hduoc2003) [reply docs/dark.md:26]: You mean I should delete this line

File: docs/dark.md [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+   is odd.
+2. There exists a good probability that $p$ is a prime number of the form $4k+1$.
+   Using [Fermat's theorem on sums of two squares](https://en.wikipedia.org/wiki/Fermat%27s_theorem_on_sums_of_two_squares), we can
+   find $\omega_{3},\omega_{4}$ efficiently. There are two stages: (i) find $z$ with $z^{2}\equiv -1 \pmod{p}$, and (ii) use $z$ to
+   find $\omega_3$ and $\omega_4$.
+
+   (i). Find $a$ where $a^{(p-1) / 2}\equiv-1\pmod{p}$, then $z=a^{(p-1)/4}$. There are over half of all number $a$ such that $1 < a <
+   p-1$, so taking $a$ at random.
+
+   (ii). Now, we set $\omega_{3}=p \mod z$ and $\omega_{4}=z\mod \omega_3$. If $\omega_{3}^{2}+\omega_{4}^{2} \neq p$, $p$ is not a
```
Comment (zk-steve) [inline docs/dark.md:73]: use pmod and \equiv

File: docs/dark.md [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+   prime number,
+   so repeat our algorithm again.
+
+For example, consider $\mu=102$. We set $\omega_{1}=1, \omega_{2}=2$, resulting in $p=97$. Choosing $a=13$, we get $z=22$. Now,
```
Comment (zk-steve) [inline docs/dark.md:77]: pmod

File: docs/dark.md (hunk 9 above)
Comment (hduoc2003) [reply docs/dark.md:73]: I believe using $mod$ is more appropriate because I want to calculate the value of $\omega$

**Comment by zk-steve** [comment]
Fill the checklist

**Pull request merged**